### ISSUE TITLE
Epic slowdown when using Bluetooth controller (MacOS)

### ISSUE THREAD

**Issue by binarize-old** [byuu/bsnes#178]
**TLDR**

Games run super slow when a Bluetooth controller is attached.

**How to reproduce**

Using a SN30 Pro Plus (firmware 3.01) with MacOS (Catalina). Controller is shown as a Dualshock 4 in the Bluetooth preference and seems to be working as expected on other games/software with no performance impact.

However, when running the bsnes 111.9 nightly (latest as of the time of this ticket), games drop from full-speed to sub 1fps when the Bluetooth device is attached. If I disable Bluetooth, the game speeds right back up to 60hz.

The controller has been mapped to usual buttons (so nothing mapped to the triggers, and none of the autofire buttons mapped). The mapping screen itself seems to work just fine with no perceivable lag when assigning buttons etc.

**Comment by byuu** [comment]
I'll see if I can track down a Bluetooth gamepad over here, but right now I don't have one to test with.

@Sintendo I don't suppose you'd have any ideas about this? It's really not my area of expertise.

**Comment by Sintendo** [comment]
I'm afraid I don't. Tried the 111.9 nightly build (as opposed to a locally compiled one) with my Switch Pro controller via Bluetooth again to be sure and that still worked fine.

Still running Mojave myself, so either Catalina broke it or the controller's doing something bonkers (repeatedly connecting/disconnecting or perhaps reporting a ridiculous number of distinct elements which slows down polling). I don't own this particular controller myself, so I can't investigate.

@welsby 
1. Have you tried any other controllers (USB or Bluetooth)?
2. Are you using the controller in Mac mode (that's a thing according to 8bitdo's support website)?
3. Are you capable and willing to compile and run some code on your machine?

**Comment by welsby** [comment]
Heya,

1.)  Just tried a Dualshock 4 (with no SN30 Pro+ also connected) and had the same issue, the moment i turn bluetooth off it jumps back to normal speed, but with it on (and a pad connected) it drops to sub 1fps.
2.) Mac mode
3.) Of course, happy to help.

Now what is interesting, is if I go to 'Controller Port 1' and select 'None' it speeds up a bit, from 1fps to around 10fps. Not sure if that helps shed some light on what the cause could be.

Thanks!

**Comment by Screwtapello** [comment]
"Siggi" popped into the #bsnes channel on Discord to comment on this issue:

> I've encountered this with osx catalina as well. I've noticed that a Switch controller works correctly, while the 8bitdo SN30 in "mac mode" does not. Not sure what causes this, but a work around is to put the SN30 in "Switch Mode".

**Comment by byuu** [comment]
In any case, this will need to be fixed by someone else. If it gets no traction I will have to close it out and default the macOS port to no joypad support.

**Comment by Sintendo** [comment]
I've found a friend of mine willing to borrow me a DualShock 4 controller. Whether I'll be able to reproduce and fix this remains to be seen, of course.

**Comment by Sintendo** [comment]
@binarize-old I've just pushed a branch with some performance logging added. Follow the instructions below to build and run bsnes. Ideally you start with your controller disconnected, then try to reproduce the issue by connecting it. After a few minutes you can close bsnes and share `output.txt` with us.

```
git clone --single-branch --branch ruby-iokit-perf https://github.com/Sintendo/bsnes.git
cd bsnes
make -C bsnes -j3 build=release target=bsnes
./bsnes/out/bsnes.app/Contents/MacOS/bsnes > output.txt
```

**Comment by binarize-old** [comment]
Thanks. built it - it's midnight over here so will run it and report back tomorrow.

**Comment by binarize-old** [comment]
Hit an unexpected issue, that debug build seems to crash out straight away:

 1467 segmentation fault  ./bsnes/out/bsnes.app/Contents/MacOS/bsnes > output.txt

Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   libdyld.dylib                 	0x00007fff71ac7476 stack_not_16_byte_aligned_error + 0
1   ???                           	0x00007ffee0404a70 0 + 140732660730480
2   org.byuu.bsnes                	0x000000010fa24e99 main + 9
3   libdyld.dylib                 	0x00007fff71ad6405 start + 1

Thread 1:
0   libsystem_pthread.dylib       	0x00007fff71cdf5b4 start_wqthread + 0

Thread 2:
0   libsystem_pthread.dylib       	0x00007fff71cdf5b4 start_wqthread + 0

Thread 3:
0   libsystem_pthread.dylib       	0x00007fff71cdf5b4 start_wqthread + 0

Thread 4:
0   libsystem_pthread.dylib       	0x00007fff71cdf5b4 start_wqthread + 0

**Comment by Sintendo** [comment]
I've pushed some changes that should at least fix the segmentation fault. Would you mind trying again?

**Comment by kode54** [comment]
I can try too, as I too have a DualShock 4 pad, and run Catalina.

**Comment by kode54** [comment]
What the bloody hell?

```
appendJoypad axes 553
appendJoypad buttons 14
appendJoypad hats 1
```

**Comment by kode54** [comment]
Fixed with a pull to Sintendo fork, which can be distilled into non-logging versions. Basically, these DualShock 4 pads have several hundred vendor extensions that aren't really axes, and that is determined by checking the UsagePage field, which won't be either GenericDesktop or Simulation, but instead start at the 0xFF00 offset:

```
    kHIDPage_VendorDefinedStart     = 0xFF00
```

**Comment by byuu** [comment]
@kode54 it's a public Discord, you're welcome. https://discord.gg/BpB6nZC

**Comment by byuu** [comment]
I merged @kode54's patch (just the page check and not the print statements of course.)
Will be in the next WIP, many thanks!
